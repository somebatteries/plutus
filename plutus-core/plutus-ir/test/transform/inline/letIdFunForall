{- 
A single occurrence polymorphic identity function gets inlined by the unconditional inlining
optimization.

let
        idFun :: forall a. a -> a
        idFun x = x
        g :: (Integer -> Integer) -> (Integer -> Integer)
        g y = idFun
    in g idFun 1
-}

(let
  (nonrec)

  (termbind
    (strict)
    (vardecl idFun (all a (type) (fun a a)))
    (abs a (type) (lam x a x))
  )

  (termbind
    (strict)
    (vardecl
      g
      (fun
        (fun (con integer) (con integer)) (fun (con integer) (con integer))
      )
    )
    (lam
      y
      (fun (con integer) (con integer))
      (idFun)
    )
  )
  
  [ [ g { idFun (con integer) } ] (con integer 1) ]
)