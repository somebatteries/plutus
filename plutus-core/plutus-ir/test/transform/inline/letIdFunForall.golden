(let
  (nonrec)
  (termbind
    (strict) (vardecl idFun (all a (type) (fun a a))) (abs a (type) (lam x a x))
  )
  [
    [ (lam y (fun (con integer) (con integer)) idFun) { idFun (con integer) } ]
    (con integer 1)
  ]
)